services/galley/src/Galley/API/MLS/Message.hs

```diff
@@ -660,25 +700,57 @@ processCommitWithAction qusr senderClient con lconv cm epoch groupId action send
               throw (InternalErrorWithDescription "Unexpected creator client set")
             -- the sender of the first commit must be a member
             _ -> throw (mlsProtocolError "Unexpected sender")
-          pure $ pure () -- no key package ref update necessary
+          pure $ (pure (), action) -- no key package ref update necessary
         else case (sender, upLeaf <$> cPath commit) of
           (MemberSender senderRef, Just updatedKeyPackage) -> do
             updatedRef <- kpRef' updatedKeyPackage & note (mlsProtocolError "Could not compute key package ref")
             -- postpone key package ref update until other checks/processing passed
             case senderClient of
-              Just cli -> pure $ updateKeyPackageMapping lconv qusr cli (Just senderRef) updatedRef
-              Nothing -> pure $ pure ()
-          (_, Nothing) -> pure $ pure () -- ignore commits without update path
+              Just cli -> pure (updateKeyPackageMapping lconv qusr cli (Just senderRef) updatedRef, action)
+              Nothing -> pure (pure (), action)
+          (_, Nothing) -> pure (pure (), action) -- ignore commits without update path
+          (NewMemberSender, Just newKeyPackage) -> do
+            -- this is an external commit
+            when (paExternalInit action == mempty) $
+              throwS @'MLSMissingExternalInit
+            unless (paAdd action == mempty) $
+              throwS @'MLSExternalCommitRemoveProposal
+
+            cid <- case kpIdentity (rmValue newKeyPackage) of
+              Left e -> throw (mlsProtocolError $ "Failed to parse the client identity: " <> e)
+              Right v -> pure v
+            newRef <-
+              kpRef' newKeyPackage
+                & note (mlsProtocolError "An invalid key package in the update path")
+
+            -- check if there is a key package ref in the remove proposal
+            remRef <-
+              if Map.null (paRemove action)
+                then pure Nothing
+                else do
+                  (remCid, r) <- derefUser (paRemove action) qusr
+                  unless (cidQualifiedUser cid == cidQualifiedUser remCid) $
+                    throwS @'MLSRemovalUserMismatch
+                  pure (Just r)
+
+            updateKeyPackageMapping lconv qusr (ciClient cid) remRef newRef
```

Comment:
Shouldn't you validate the key package first? See for example `applyProposal`.

Reply:
Your observation has seemed valid and with that in mind, I looked at other cases when a key package is updated or mapped and they some of them seemed equally wrong because they also just call `updateKeyPackageMapping`. For example, in the case of an existing member updating their key package via the update path we blindly accept the new key package without checking if it is for the same protocol, same cipher suite, etc., but this seems wrong, doesn't it?

I realized `updateKeyPackageMapping` does validation (the ciphersuite, identity matching, signature scheme, etc.) if the old key package is non-existent (i.e., we pass in `Nothing` to `updateKeyPackageMapping`), by calling an internal endpoint in Brig, for which there's a handler `upsertKeyPackage`.

So before making a call to `updateKeyPackageMapping` in the case of a `NewMemberSender`, I just added a call to `addKeyPackageRef`, which calls `upsertKeyPackage` in Brig and validates it before inserting into the database.

Does that seem OK?

Reply:
As long as validation is happening.